### ISSUE TITLE
Group_number issue with attention layers

### ISSUE THREAD

**Issue by charliezjw** [sIncerass/QBERT#2]
Thank you for your previous help. One more difficulty that I encountered:
I am trying to add group_number to the weight matrix of query, key, and values. I am doing so by adding group_number parameter to [here](https://github.com/sIncerass/QBERT/blob/cb8fe57dc078582a49bd4081ded220241187c6be/modeling.py#L144  ) . But it does not take any effect. I tried adding group_number to [BertIntermediate_Quant](https://github.com/sIncerass/QBERT/blob/cb8fe57dc078582a49bd4081ded220241187c6be/modeling.py#L219) as well, but it did not work.

The group_number parameter only works on [BertOutput_Quant](https://github.com/sIncerass/QBERT/blob/cb8fe57dc078582a49bd4081ded220241187c6be/modeling.py#L236)

I think I am making some mistake. Can you help me with that?

Thank you very much!

Jiawei

**Comment by sIncerass** [comment]
Hi @charliezjw, have you tried various group number setting?

**Comment by charliezjw** [comment]
I have tried 8, 12, 24 for group_number none of them worked.

Here is what I did:
Add group_number = 24 to [attention here](https://github.com/sIncerass/QBERT/blob/cb8fe57dc078582a49bd4081ded220241187c6be/modeling.py#L144) on line 144, 145, and 146. 

[Here](https://github.com/sIncerass/QBERT/blob/cb8fe57dc078582a49bd4081ded220241187c6be/quant_modules.py#L321) add:
        if self.group_number != 1:
            print(self.group_number)
But it never gets triggered.

However, if I add the group_number to [the output here](https://github.com/sIncerass/QBERT/blob/cb8fe57dc078582a49bd4081ded220241187c6be/modeling.py#L236), the print statement can be properly triggered. Or if change the [default initialization of group number here](https://github.com/sIncerass/QBERT/blob/cb8fe57dc078582a49bd4081ded220241187c6be/quant_modules.py#L262) to any number, it works for all operations, including attention as well.

So it feels like the group_number variable is touched somewhere for the attention module, but I cannot find it. Or something else is wrong. Really need some help here, have been debugging for a few day 😢 

Thank you very much!!
Jiawei

**Comment by sIncerass** [comment]
hi @charliezjw , if you setting other parameters like weight_bits / hidden_size, will it take any effect?

**Comment by charliezjw** [comment]
I tried to modify weight_bits and it worked. I did so by changing the value [here in schedule_blending_runner.py](https://github.com/sIncerass/QBERT/blob/cb8fe57dc078582a49bd4081ded220241187c6be/schedule_blending_runner.py#L15).

I am wondering whether there is any script/code I can use to reproduce the section 3.2 Effects of group-wise quantization in your q-bert paper. I am just tucked, I am hoping to get a starting point, and hopefully get the rest myself. I am hoping to start by reproducing your result.

Thank you very much!!!
